### PR TITLE
Migrate stuck invites

### PR DESCRIPTION
**Pull request by kegsay**
With tests.

### CODE REVIEW

File: state/migrations/20231108122539_clear_stuck_invites.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	// room into rooms.join"? For every room in rooms.join, we should have seen a create event for it, and hence have
+	// an entry in syncv3_rooms. If we do not have an entry in syncv3_rooms but do have an entry in syncv3_unread, this
+	// implies we failed to properly store this joined room and therefore the user who the unread marker is for should be
+	// reset to force an initial sync. On matrix.org, of the users using sliding sync, this will catch around ~1.82% of users
+	rows, err := tx.QueryContext(ctx, `
+		SELECT distinct(user_id) FROM syncv3_unread
+		WHERE room_id NOT IN (
+			SELECT room_id
+			FROM syncv3_rooms
+		) GROUP BY user_id
```
Comment (DMRobertson) [inline state/migrations/20231108122539_clear_stuck_invites.go:36]: Is the GROUP BY needed here, given that we're already using DISTINCT?

File: state/migrations/20231108122539_clear_stuck_invites_test.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	// Test cases:
+	// Room | In Invite Table? | In Unread Table? | In Room Table? | Comment
+	//   A           Y                 Y                Y             OK, Genuine invite, proxy in room
+	//   B           Y                 Y                N             BAD, Stuck invite, proxy never saw join
+	//   C           Y                 N                Y             OK, Genuine invite, proxy in room, no unread counts (unusual but valid)
+	//   D           Y                 N                N             OK, Genuine invite, proxy not in room
+	//   E           N                 Y                Y             OK, Genuine joined room
+	//   F           N                 Y                N             BAD, Stuck joined room, proxy never saw join
+	//   G           N                 N                Y             OK, Genuine joined room, no unread counts (unusual but valid)
+	//   -           N                 N                N             Impossible, room id isn't in any table!
```
Comment (DMRobertson) [inline state/migrations/20231108122539_clear_stuck_invites_test.go:52]: There's an assumption here that we never delete from the rooms table. Is that true? (I think it should be, but let's check)

File: state/migrations/20231108122539_clear_stuck_invites_test.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	// invites for Bob and Doris are gone
+	for _, userID := range []string{bob, doris} {
+		got, err := inviteTable.SelectAllInvitesForUser(userID)
+		if err != nil {
+			t.Fatalf("SelectAllInvitesForUser: %s", err)
+		}
+		if len(got) > 0 {
+			t.Fatalf("SelectAllInvitesForUser got invites for %s, wanted none: %+v", userID, got)
+		}
+	}
```
Comment (DMRobertson) [inline state/migrations/20231108122539_clear_stuck_invites_test.go:211]: Can we check that the genuine invites remain too?

**Review by DMRobertson — APPROVED** [review]

File: state/migrations/20231108122539_clear_stuck_invites_test.go (hunk 2 above)
Comment (kegsay) [reply state/migrations/20231108122539_clear_stuck_invites_test.go:52]: Yes, with the exception of a migration which did delete "bad" rooms.

File: state/migrations/20231108122539_clear_stuck_invites.go (hunk 1 above)
Comment (kegsay) [reply state/migrations/20231108122539_clear_stuck_invites.go:36]: No it isn't.

**Pull request merged**